File: sourceFile/Cpp/11_ContainerWithMostWater.cpp
/*
 * @Author: AlanGolphi
 * @Date: 2020-07-18 16:56:16
 * @LastEditTime: 2020-08-16 15:51:47
 */
class Solution {
public:
    int maxArea(vector<int>& height)
    {
        int mos = 0;
        int i = 0;
        int j = height.size() - 1;
        while (i < j) {
            int water = (j - i) * min(height[i], height[j]);
            mos = max(mos, water);
            if (height[i] < height[j]) {
                i++;
            } else
                j--;
        }
        return mos;
    }
};
//双指针法，i 和 j 初始是分别是数组的两端
//求 water 时，取数组里较小的数，之后每次把较大的赋值给 mos

File: sourceFile/typescript/11_ContainerWIthMostWater.ts
// @algorithm @lc id=11 lang=typescript
// @title container-with-most-water
// @test([1,8,6,2,5,4,8,3,7])=49
// @test([1,1])=1
function maxArea(height: number[]): number {
  let max: number = 0;
  let left = 0;
  let right = height.length - 1;

  while (left < right) {
    const leftVal = height[left];
    const rightVal = height[right];
    const innerMax = Math.min(leftVal, rightVal) * (right - left);
    max = Math.max(max, innerMax);

    if (leftVal < rightVal) {
      left++;
    } else {
      right--;
    }
  }

  return max;
}

console.log("1: ", maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]));

console.log("2: ", maxArea([1, 1]));
